### ISSUE TITLE
Edge agent doesn't support docker restart policies

### ISSUE THREAD

**Issue by uecasm** [Azure/iotedge#6992]
## Expected Behavior
My main goal is to be able to issue a `docker stop <container>` command on the edge host for a container created by iotedge and have it remain stopped until I manually start it again, while still having it automatically restart if it crashes unexpectedly.

## Current Behavior
When using `"restartPolicy": "always"` in the `$edgeAgent` modules top level desired properties, the module is automatically restarted even if manually stopped.

It's not possible to specify `"unless-stopped"` at the top level; that's the subject of #6768 and https://feedback.azure.com/d365community/idea/d1aa4fc2-8ec4-ed11-a81b-000d3ae6a6aa

----

When attempting to use the `RestartPolicy` setting within the docker `createOptions`, the agent throws an exception and fails to start the module.

(At the docker level, a restart policy of either `"always"` or `"unless-stopped"` should result in automatic restarts when not stopped manually -- and indeed this does work as expected when configured manually; it just doesn't seem to work via iotedge.)

## Steps to Reproduce
For the top level restart policy, configure a deployment manifest with content similar to:

```json
{
    "modulesContent": {
        "$edgeAgent": {
            "properties.desired.modules.foo": {
                "settings": {
                    "image": "some_image_path"
                },
                "type": "docker",
                "status": "running",
                "restartPolicy": "always",
                "version": "1.0"
            }
        }
    }
}
```

1. Deploy.
2. `docker ps` and observe that the module is running.
3. `docker stop foo`
4. `docker ps` periodically and observe that that module is initially stopped but after a short while will be running again.

----

For the internal docker policy, configure similar to:

```json
{
    "modulesContent": {
        "$edgeAgent": {
            "properties.desired.modules.foo": {
                "settings": {
                    "image": "some_image_path",
                    "createOptions": "{\"HostConfig\":{\"RestartPolicy\":{\"Name\":\"always\"}}}"
                },
                "type": "docker",
                "status": "running",
                "restartPolicy": "never",
                "version": "1.0"
            }
        }
    }
}
```

1. Attempt to deploy.
2. Observe module not running and `$edgeAgent` reporting error 500.  Error logs given below.

### Output of `iotedge check`

```
Configuration checks (aziot-identity-service)
---------------------------------------------
√ keyd configuration is well-formed - OK
√ certd configuration is well-formed - OK
√ tpmd configuration is well-formed - OK
√ identityd configuration is well-formed - OK
√ daemon configurations up-to-date with config.toml - OK
√ identityd config toml file specifies a valid hostname - OK
√ aziot-identity-service package is up-to-date - OK
√ host time is close to reference time - OK
√ preloaded certificates are valid - OK
√ keyd is running - OK
√ certd is running - OK
√ identityd is running - OK
√ read all preloaded certificates from the Certificates Service - OK
√ read all preloaded key pairs from the Keys Service - OK
√ check all EST server URLs utilize HTTPS - OK
√ ensure all preloaded certificates match preloaded private keys with the same ID - OK

Connectivity checks (aziot-identity-service)
--------------------------------------------
√ host can connect to and perform TLS handshake with iothub AMQP port - OK
√ host can connect to and perform TLS handshake with iothub HTTPS / WebSockets port - OK
√ host can connect to and perform TLS handshake with iothub MQTT port - OK

Configuration checks
--------------------
√ aziot-edged configuration is well-formed - OK
√ configuration up-to-date with config.toml - OK
√ container engine is installed and functional - OK
√ configuration has correct URIs for daemon mgmt endpoint - OK
√ aziot-edge package is up-to-date - OK
√ container time is close to host time - OK
√ DNS server - OK
√ production readiness: logs policy - OK
√ production readiness: Edge Agent's storage directory is persisted on the host filesystem - OK
√ production readiness: Edge Hub's storage directory is persisted on the host filesystem - OK
√ Agent image is valid and can be pulled from upstream - OK
√ proxy settings are consistent in aziot-edged, aziot-identityd, moby daemon and config.toml - OK

Connectivity checks
-------------------
√ container on the default network can connect to upstream AMQP port - OK
√ container on the default network can connect to upstream HTTPS / WebSockets port - OK
√ container on the IoT Edge module network can connect to upstream AMQP port - OK
√ container on the IoT Edge module network can connect to upstream HTTPS / WebSockets port - OK
35 check(s) succeeded.
2 check(s) were skipped due to errors from other checks. Re-run with --verbose for more details.
```

### Device Information
* Host OS [e.g. Ubuntu 18.04, Windows Server IoT 2019]: Ubuntu 20.04.6 LTS
* Architecture [e.g. amd64, arm32, arm64]: amd64
* Container OS [e.g. Linux containers, Windows containers]: Linux containers

### Runtime Versions
* aziot-edged [run `iotedge version`]: iotedge 1.4.9
* Edge Agent [image tag (e.g. 1.0.0)]:  1.4.0
* Edge Hub [image tag (e.g. 1.0.0)]:  1.4.0
* Docker/Moby [run `docker version`]: 20.10.23+azure-2

## Logs

Edge agent log for internal restart policy configuration:

```
<6> 2023-04-11 05:16:05.774 +00:00 [INF] - Executing command: "Command Group: (\n  [Command Group: (\n  [Prepare module foo]\n  [Create module foo]\n)]\n  [Start module foo]\n)"
<6> 2023-04-11 05:16:05.775 +00:00 [INF] - Executing command: "Command Group: (\n  [Prepare module foo]\n  [Create module foo]\n)"
<3> 2023-04-11 05:16:39.753 +00:00 [ERR] - Executing command for operation ["create"] failed.
Microsoft.Azure.Devices.Edge.Agent.Edgelet.EdgeletCommunicationException- Message:Error calling Create module foo: runtime operation error: create module "foo", StatusCode:500, at:   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2022_08_03.ModuleManagementHttpClient.HandleException(Exception exception, String operation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2022_08_03/ModuleManagementHttpClient.cs:line 232
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/versioning/ModuleManagementHttpClientVersioned.cs:line 167
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2022_08_03.ModuleManagementHttpClient.CreateModuleAsync(ModuleSpec moduleSpec) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2022_08_03/ModuleManagementHttpClient.cs:line 102
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.<>c__DisplayClass26_0.<<Throttle>b__0>d.MoveNext() in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 145
--- End of stack trace from previous location ---
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.Throttle[T](Func`1 identityOperation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 164
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 69
<3> 2023-04-11 05:16:39.753 +00:00 [ERR] - Executing command for operation ["Command Group: (\n  [Command Group: (\n  [Prepare module foo]\n  [Create module foo]\n)]\n  [Start module foo]\n)"] failed.
Microsoft.Azure.Devices.Edge.Agent.Edgelet.EdgeletCommunicationException- Message:Error calling Create module foo: runtime operation error: create module "foo", StatusCode:500, at:   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2022_08_03.ModuleManagementHttpClient.HandleException(Exception exception, String operation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2022_08_03/ModuleManagementHttpClient.cs:line 232
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/versioning/ModuleManagementHttpClientVersioned.cs:line 167
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2022_08_03.ModuleManagementHttpClient.CreateModuleAsync(ModuleSpec moduleSpec) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2022_08_03/ModuleManagementHttpClient.cs:line 102
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.<>c__DisplayClass26_0.<<Throttle>b__0>d.MoveNext() in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 145
--- End of stack trace from previous location ---
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.ModuleManagementHttpClient.Throttle[T](Func`1 identityOperation) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/ModuleManagementHttpClient.cs:line 164
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 69
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /mnt/vss/_work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 69
<3> 2023-04-11 05:16:39.754 +00:00 [ERR] - Step failed in deployment 18, continuing execution. Failure when running command Command Group: (
  [Command Group: (
  [Prepare module foo]
  [Create module foo]
)]
  [Start module foo]
). Will retry in -07s.
```

Docker log (different time, same error):
```
Apr 11 16:48:32 insight-edge dockerd[841]: time="2023-04-11T16:48:32.847519925+12:00" level=error msg="Handler for POST /containers/create returned error: json: cannot unmarshal number into Go struct field RestartPolicy.HostConfig.RestartPolicy.Name of type string"
```

## Additional Information
This looks like a bug in your API serialization; you're apparently parsing to an enum and then passing that back out as an integer instead of a string.  idk why you'd even be converting this in the first place instead of just passing it verbatim; isn't that the point of passing the docker config stringified instead of directly?

**Comment by vipeller** [comment]
Hi @uecasm,

I believe that the problem is that the internal structure of edgeAgent does not have a value for unless-stopped:

https://github.com/Azure/iotedge/blob/main/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/IModule.cs#L74

Regarding the question that why it does not keep the string, edgeAgent originally was not designed to work only on docker, and although its configuration is basically mirroring docker, it still has its own internal structure and also has e.g. its own restart logic implemented around the restart policy values. That is why the code of IModule has an "on-unhealthy" , which I suppose was created when edge still supported kubernetes directly. That time it was not planned to support "everything" docker supports, but more like a mix of cases different systems supported that time.

Anyway, today edge works only with docker and the case you want to implement makes sense. Let me play a bit with the source to see how that value can go through.

If it is not a big change, we can add it, but be aware that due to the release loop, it can take several weeks or 1-2 months until it gets released.

**Comment by bishal41** [comment]
Hi @uecasm I hope the explanation from @vipeller was helpful. Also, thanks for submitting the feedback and including additional details. We have your feedback in our backlog. We are considering IoT Edge 1.4 (LTS) as a feature complete.  Due to the resource constraints, we will consider adding any additional features in next 9-12 months. I just want to set the right expectations with you. Please let me know if you have any questions and whether this issue can be resolved? Thank you!

**Comment by uecasm** [comment]
This is not a feature request, this is a bug report.  As such, I would hope that you would fix it even for 1.4, unless you're completely locked even for bugfixes (which would be unfortunate).

Regarding @vipeller's answer, it's interesting but not directly relevant.  As noted, even the `always` value (which is already listed in that enum) does not work, because it sends a number to the API and not the string.  The problem is lower level than that; you haven't configured your JSON serialization options correctly.  (Unless this was already fixed in a later version than I have installed, or the error lies in a different part of the code.)

Support for other values such as `unless-stopped` would be nice (and that would be a feature request), but that was not what I was reporting.

**Comment by uecasm** [comment]
Specifically, the bug is that you have not set the string-enum converter on **this** class:

https://github.com/Azure/iotedge/blob/main/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/models/RestartPolicyKind.cs

**Comment by vipeller** [comment]
@uecasm  sorry my attention got stuck on 'unless-stopped' and did not focus on your example later. I see what you are saying and thanks for your analysis, it is something easy to fix and we should do so.

**Comment by uecasm** [comment]
So, I see #7006 has been merged to `main`, but the history of `main` seems unrelated to what's actually getting released for 1.4; as far as I can tell it's not yet fixed in any 1.4 release?  Any chance that can be corrected?  This is quite an aggravating bug for my use-case, and it seems a very straightforward fix.

**Comment by vipeller** [comment]
@uecasm Hi, now the change is flagged to be released with the next 1.4 release. Here is the PR on 1.4: https://github.com/Azure/iotedge/pull/7101. Sorry for the delay.
